src/notification.c

```diff
@@ -280,55 +283,96 @@ char *notification_extract_markup_urls(char **str_ptr) {
+ *
+ * This function is guaranteed to return a valid pointer.
+ * @Returns: The generated notification
  */
 notification *notification_create(void)
 {
-        return g_malloc0(sizeof(notification));
-}
+        //FIXME: handle OOM
+        notification *n = g_malloc0(sizeof(notification));
 
-void notification_init_defaults(notification *n)
-{
-        assert(n != NULL);
-        if(n->appname == NULL) n->appname = g_strdup("unknown");
-        if(n->summary == NULL) n->summary = g_strdup("");
-        if(n->body == NULL) n->body = g_strdup("");
-        if(n->category == NULL) n->category = g_strdup("");
+        /* Unparameterized default values */
+        n->first_render = true;
+        n->markup = settings.markup;
+        n->format = settings.format;
+
+        n->timestamp = g_get_monotonic_time();
+
+        n->urgency = NORM;
+        n->timeout = -1;
+
+        n->transient = false;
+        n->progress = -1;
+
+        return n;
 }
 
 /*
- * Initialize the given notification
+ * Sanitize values of notification, apply all matching rules
+ * and generate derived fields.
+ *
+ * TODO: (?) This should be callable all the time
+ *           maybe we should split the function calls for UPDATE derived fields into a separate calling function
  *
- * n should be a pointer to a notification allocated with
- * notification_create, it is undefined behaviour to pass a notification
- * allocated some other way.
+ * @n: the notification to sanitize
  */
 void notification_init(notification *n)
 {
-        assert(n != NULL);
-
-        //Prevent undefined behaviour by initialising required fields
-        notification_init_defaults(n);
-
-        n->script = NULL;
-        n->text_to_render = NULL;
-
-        n->format = settings.format;
-
-        rule_apply_all(n);
-
+        /* default to empty string to avoid further NULL faults */
+        n->appname  = n->appname  ? n->appname  : g_strdup("unknown");
+        n->summary  = n->summary  ? n->summary  : g_strdup("");
+        n->body     = n->body     ? n->body     : g_strdup("");
+        n->category = n->category ? n->category : g_strdup("");
+
+        /* sanitize urgency */
+        if (n->urgency < LOW)
+                n->urgency = LOW;
+        else if (n->urgency > CRIT)
+                n->urgency = CRIT;
+
+        /* Timeout processing */
+        if (n->timeout < 0)
+                n->timeout = settings.timeouts[n->urgency];
+
+        /* Icon handling */
         if (n->icon != NULL && strlen(n->icon) <= 0) {
                 g_free(n->icon);
                 n->icon = NULL;
         }
-
         if (n->raw_icon == NULL && n->icon == NULL) {
                 n->icon = g_strdup(settings.icons[n->urgency]);
         }
 
-        n->urls = notification_extract_markup_urls(&(n->body));
+        /* Color hints */
+        n->colors[ColFG] = n->colors[ColFG]
+                         ? n->colors[ColFG]
+                         : xctx.colors[ColFG][n->urgency];
```

Comment:
I *really* don't like using `xctx` inside non-X related functions. Though I guess this will have to stay as-is until the drawing refactor is done.

Reply:
I agree with you. But previously, `notification_init` also had to access this struct. It's bad, but it's not getting worse (and hopefully gets fixed in refactor drawing).

Reply:
Also, I can't help but feel like all these ternary operators are a bit ugly - what's wrong with a usual `if`?